### PR TITLE
feat(ctb): Refactor `SemverLock`

### PR DESCRIPTION
**Pull request by clabby**
## Overview

Breaks out the semver lock hash in `contracts-bedrock` into two separate hashes: the sourcecode hash & the initcode hash.

By the [style guide](https://github.com/ethereum-optimism/optimism/blob/develop/packages/contracts-bedrock/STYLE_GUIDE.md#versioning):
1. A patch bump should occur when the source changes, but the initcode hash does not.
2. A minor bump should occur when the source + initcode hash changes, but there are no breaking changes to the interface.
3. A major bump should occur when there are breaking changes to the interface or large changes to the security model.

This extra granularity allows for reviewers to more easily assess whether or not the semver bump in smart contract PRs is appropriate for the changes. Previously, the single hash of `sourceCode ++ initCode` required a more in-depth analysis of the changes to determine if the bump was appropriate.

### CODE REVIEW

**Comment by clabby** [comment]
Current dependencies on/for this PR:
* develop
  * **PR #8390** <a href="https://app.graphite.dev/github/pr/ethereum-optimism/optimism/8390?utm_source=stack-comment-icon" target="_blank">Graphite</a>  👈

This [stack of pull requests](https://stacking.dev/?utm_source=stack-comment) is managed by [Graphite](https://app.graphite.dev/github/pr/ethereum-optimism/optimism/8390?utm_source=stack-comment).

**Review by tynes — APPROVED** [review]
Looks good to me, json in solidity is so weird

File: packages/contracts-bedrock/scripts/SemverLock.s.sol [hunk 1]
```diff
@@ -53,13 +53,12 @@ contract SemverLock is Script {
...
+            // Serialize the initcode hash + sourcecode hash in JSON.
+            vm.serializeBytes32(_files[i], "initCodeHash", keccak256(initCode));
+            string memory obj = vm.serializeBytes32(_files[i], "sourceCodeHash", keccak256(bytes(fileContents)));
 
-            // If this is the last file, set the output.
-            if (i == _files.length - 1) {
-                out = j;
-            }
+            // Serialize the map from the file name -> initcode hash + sourcecode hash in JSON.
+            out = vm.serializeString("", _files[i], obj);
```
[the remark this answers is not in the document]
Comment (refcell) [reply packages/contracts-bedrock/scripts/SemverLock.s.sol:61]: Thanks for the tip rabbit, but `vm.serializeString` tracks the serialized object by the empty string key, so overwriting the `out` variable doesn't matter in this case

**Review by refcell — APPROVED** [review]

File: packages/contracts-bedrock/scripts/SemverLock.s.sol (hunk 1 above)
[the remark this answers is not in the document]
Comment (tynes) [reply packages/contracts-bedrock/scripts/SemverLock.s.sol:61]: You should read the foundrybook and explain to me how json serialization works. Your comment is not correct given the outputs of the script in `semver-lock.json`.

**Comment by tynes** [comment]
@clabby I rebased this to fix the merge conflict

**Pull request merged**